### PR TITLE
Migrate to Async

### PR DESCRIPTION
**Pull request by thearchitector**
Relates to #55 

## Changes

- All methods are now async/await coroutines.
- Synchronous routes are run in a threadpool from an async wrapper, like how [Starlette does it](https://github.com/encode/starlette/blob/537ab6afd110b79dca95f3c8ecc6980710b1de1c/starlette/routing.py#L51-L66).
- Blocking calls to `limits` storage backends or limiters is also executed in a threadpool.

**NOTE**: Decorated functions now always become async coroutines, which should not cause any issues unless developers are invoking their routes directly.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -14,16 +14,17 @@ documentation = "https://slowapi.readthedocs.io/en/latest/"
...
 redis = "^3.4.1"
+anyio = "3.0.0"
 
 [tool.poetry.dev-dependencies]
 isort = "^4.3.21"
 mypy = "^0.761"
 black = "^19.10b0"
 fastapi = "^0.61.0"
-lxml = "^4.5.0"
+lxml = { version = "^4.5.0", markers="platform_system != 'Windows'" }
```
Comment (thearchitector) [inline pyproject.toml:27]: I actually develop on Windows, which caused problems with this dependency. Whether or not it's included is kinda up to you, but I don't see any reason not to include the environment marker. Just means people on Windows can't build the documentation.

File: pyproject.toml [hunk 2]
```diff
@@ -14,16 +14,17 @@ documentation = "https://slowapi.readthedocs.io/en/latest/"
 include = ["slowapi/py.typed"]
 
 [tool.poetry.dependencies]
-python = "^3.6"
+python = "^3.6.2"
 limits = "^1.5"
 redis = "^3.4.1"
+anyio = "3.0.0"
```
Comment (thearchitector) [inline pyproject.toml:20]: Also question, perhaps about python development in general, but what's the reason for not including `starlette` as a py dependency? It is definitely required, after all.

File: pyproject.toml (hunk 1 above)
Comment (laurentS) [reply pyproject.toml:27]: I'm surprised it even works on windows, to be honest :)
But agree with you, I don't think there's an issue in adding the marker, I would think that most people will simply read the docs online rather than rebuild them locally.

**Comment by laurentS** [comment]
@thearchitector Thanks for putting this together, it's definitely a great step forward!
I only have 2 concerns:
- how do we move to using async limits? I've not seen anything happen on that PR for over a year now, it might be worth considering releasing that code as a separate package, something like `async-limits` and use it here?
- your note about all routes becoming async coroutines. I'm worried that this might break things for some users. I have no way of knowing if it actually is a use case that anyone has, unfortunately. Would it be worth keeping the sync decorator for a bit longer, and issuing a deprecation warning on sync routes? Something like "Starting next release, sync routes will be turned into coroutines by the limits decorator".

Trying to think of a way forward, we could change your PR to readd the sync part (with deprecation warning) with a short explanation in the docs (maybe linked from the warning message), do a release with it, then package `async-limits`, include it in `slowapi` and release again, this time going all async.

What do you think?

File: pyproject.toml (hunk 2 above)
Comment (laurentS) [reply pyproject.toml:20]: It's a very good question, and I don't think I have a good answer to it :)
I can't see moving `starlette` up here being a problem in any way.

**Review by laurentS — COMMENTED** [review]
I messed up my review, sorry.
See my other comments on the PR :)

**Comment by laurentS** [comment]
relates to #3

**Comment by laurentS** [comment]
@thearchitector any thoughts on my comments above?

**Comment by thearchitector** [comment]
@laurentS sorry, i've been moving around a lot and haven't gotten a chance to get back to this. but its on my todos! re ideas, i agree -- its not looking likely that limits is going to implement the async stuff, so i think it makes sense to fork it. tbh and imo its much cleaner with a fork anyways

about the routes, yea... it bothered me enough when i did it to mention it in the PR desc, so idk how happy i am about it afterall. the logic was that any other libraries using the routes directly would have to check if its a coroutine or not anyway, and by doing it I could avoid a lot of `event loop <--> worker thread` switching (since `limits` was blocking and needed to be dispatched). BUT thinking more, it introduces some "what you wrote is not what you get" weirdness from the developer POV that i could see being annoying. migrating to `async-limits` or whatever we call it will also fix the switching issue since it wont be blocking anymore.

tl;dr i'm for an async fork, and for long-term keeping both def and async routes with a better way of dispatching the sync ones to the event loop than i had before

**Comment by thearchitector** [comment]
@laurentS letting you know I haven't forgotton about this, just have not had a huge amount of free time to work. nonetheless, https://github.com/thearchitector/alimits/ now exists and seems to be pretty good (by the initial code you wrote in the PR and some reorganization i did for my own brain)

this PR is next
